File: MediaBase/MediaBase/RawVideoFrame.m
//
//  RawVideoFrame.m
//  MediaBase
//
//  Created by tbago on 16/12/29.
//  Copyright © 2016年 tbago. All rights reserved.
//

#import "RawVideoFrame.h"

@interface RawVideoFrame()

@property (nonatomic, readwrite) ResuablePixelFormat   pixelFormat;
@property (nonatomic, readwrite) uint32_t              width;
@property (nonatomic, readwrite) uint32_t              height;

@property (strong, nonatomic) NSMutableArray    *innerLineSizeArray;
@property (strong, nonatomic) NSMutableArray    *innerFrameDataArray;

@end

@implementation RawVideoFrame

- (instancetype)init {
    return nil;
}

- (instancetype)initWithPixFormat:(ResuablePixelFormat) pixelFormat
                            width:(uint32_t) width
                           height:(uint32_t) height
{
    self = [super init];
    if (self) {
        self.pixelFormat = pixelFormat;
        self.width = width;
        self.height = height;
    }
    return self;
}

#pragma mark - public method

- (void)pushFrameData:(uint32_t)lineSize
            frameByte:(uint8_t *)frameByte {
    
    [self.innerLineSizeArray addObject:@(lineSize)];
    
    NSInteger byteLength = [self calcFrameByteLength];
    NSData *frameData = [[NSData alloc] initWithBytes:frameByte length:byteLength];
    
    [self.innerFrameDataArray addObject:frameData];
}

#pragma mark - private method

- (NSInteger)calcFrameByteLength {
    NSInteger byteLength = 0;
    NSInteger lineIndex = self.innerLineSizeArray.count - 1;
    switch(self.pixelFormat)
    {
        case R_PIX_FMT_YUVJ420P:
        case R_PIX_FMT_YUV420P:
        case R_PIX_FMT_YUV420P10LE:
            if (lineIndex == 0) {
                byteLength = [self.innerLineSizeArray[lineIndex] integerValue] * self.height;
            }
            else if (lineIndex > 0) {
                byteLength = [self.innerLineSizeArray[lineIndex] integerValue] * (self.height>>1);
            }
            break;
        case R_PIX_FMT_RGB24:     // Packed pixel, 3 bytes per pixel, RGBRGB...
        case R_PIX_FMT_BGR24:
            byteLength = [self.innerLineSizeArray[lineIndex] integerValue] * self.height * 3;
            break;
        default:
            NSLog(@"not support pixformat");
            break;
    }
    return byteLength;
}

#pragma mark - get & set

- (NSMutableArray *)innerLineSizeArray {
    if (_innerLineSizeArray == nil) {
        _innerLineSizeArray = [[NSMutableArray alloc] init];
    }
    return _innerLineSizeArray;
}

- (NSMutableArray *)innerFrameDataArray {
    if (_innerFrameDataArray == nil) {
        _innerFrameDataArray = [[NSMutableArray alloc] init];
    }
    return _innerFrameDataArray;
}

- (NSArray *)lineSizeArray {
    return self.innerLineSizeArray;
}

- (NSArray *)frameDataArray {
    return self.innerFrameDataArray;
}

@end

File: MediaBase/MediaBase/RawVideoFrame.cc
//
//  RawVideoFrame.m
//  MediaBase
//
//  Created by tbago on 16/12/29.
//  Copyright © 2021 tbago. All rights reserved.
//

#include "RawVideoFrame.h"

namespace media_base {

RawVideoFrame::RawVideoFrame(ResuablePixelFormat pixelFormat, uint32_t width, uint32_t height)
: pixelFormat(pixelFormat),
width(width),
height(height)
{
    timeStamp = 0;
    duration = 0;
}

RawVideoFrame::RawVideoFrame(const RawVideoFrame &videoFrame)
{
    CopyInfo(videoFrame);
}

RawVideoFrame & RawVideoFrame::operator=(const RawVideoFrame &videoFrame)
{
    if (this == &videoFrame) {
        return *this;
    }
    CopyInfo(videoFrame);
    return *this;
}

RawVideoFrame::~RawVideoFrame()
{
    FreeFrameBuffer();
}

void RawVideoFrame::PushFrameData(uint32_t lineSize, int8_t *frameData) {
    RawVideoFrameBuffer *frameBuffer = new RawVideoFrameBuffer();
    frameBuffer->lineSize = lineSize;
    uint32_t lineIndex = (uint32_t)frameBufferVector.size();
    switch(pixelFormat) {
        case R_PIX_FMT_YUV420P:
            if (lineIndex == 0) {
                frameBuffer->lineCount = height;
            }
            else {
                frameBuffer->lineCount = height >> 1;
            }
            break;
        case R_PIX_FMT_RGB24:     // Packed pixel, 3 bytes per pixel, RGBRGB...
        case R_PIX_FMT_BGR24:
            frameBuffer->lineCount = height;
            break;
        default:
            assert(false);
            break;
    }

    uint32_t frameDataLength = frameBuffer->lineSize * frameBuffer->lineCount;
    frameBuffer->frameData = (int8_t *)malloc(frameDataLength);
    memcpy(frameBuffer->frameData, frameData, frameDataLength);
    frameBufferVector.push_back(frameBuffer);
}

void RawVideoFrame::CopyInfo(const RawVideoFrame &videoFrame) {
    pixelFormat = videoFrame.pixelFormat;
    width = videoFrame.width;
    height = videoFrame.height;
    timeStamp = videoFrame.timeStamp;
    duration = videoFrame.duration;

    FreeFrameBuffer();
    for (uint32_t i = 0; i < videoFrame.frameBufferVector.size(); i++) {
        RawVideoFrameBuffer *frameBuffer = videoFrame.frameBufferVector[i];
        PushFrameData(frameBuffer->lineSize, frameBuffer->frameData);
    }
}

void RawVideoFrame::FreeFrameBuffer()
{
    for (uint32_t i = 0; i < frameBufferVector.size(); i++) {
        RawVideoFrameBuffer *frameBuffer = frameBufferVector[i];
        free(frameBuffer->frameData);
        delete frameBuffer;
        frameBufferVector[i] = NULL;
    }
    frameBufferVector.clear();
}

}   //namespace media_base
